### ISSUE TITLE
CC allows illegal host names for apps

### ISSUE THREAD

**Issue by domdom82** [cloudfoundry/cloud_controller_ng#1593]
## Issue

When pushing an app, by default the app name becomes part of the FQDN of the app.
CC currently allows app/host names that don't comply to [RFC 1034](https://tools.ietf.org/html/rfc1034)
This creates host names with illegal characters, e.g. underscores `_` which leads to a whole lot of problems with DNS.

## Context

RFC 1034 states that
```
The labels must follow the rules for ARPANET host names.  They must
start with a letter, end with a letter or digit, and have as interior
characters only letters, digits, and hyphen.  There are also some
restrictions on the length.  Labels must be 63 characters or less.
```

## Steps to Reproduce

```
cf map-route my-app cfapps.mycloud.com --hostname my_app
Creating route my_app.cfapps.mycloud.com for org test / space test as cf_admin...
OK
Adding route my_app.cfapps.mycloud.com to app my-app in org test / space test as cf_admin...
OK
```
## Expected result

```
ERROR illegal hostname
```

## Current result

```
OK
```

## Possible Fix

CC should prevent the mapping of non-compliant hostnames to apps.

**Comment by cwlbraa** [comment]
This should already be implemented on V3. We would accept a PR to fix it in V2, but it's already deprecated and we're working to remove it.

**Comment by domdom82** [comment]
Hi, @cwlbraa I think the problem still exists in V3. I was able to post a route with underscores using V3 API:

```
REQUEST: [2020-04-15T09:32:35Z]
POST /v3/routes HTTP/1.1
Host: api.cf.mycloud.com
Accept: application/json
Authorization: [PRIVATE DATA HIDDEN]
Content-Type: application/json
User-Agent: cf7/7.0.0-beta.30+6203bbe5d.2020-02-21 (go1.13.4; amd64 linux)
{
  "host": "this_should_not_work",
  "relationships": {
    "domain": {
      "data": {
        "guid": "c542069e-06a6-481c-b8ba-a055bb2cd086"
      }
    },
    "space": {
      "data": {
        "guid": "c656ca01-6846-4f9d-a1e3-ff026ffaeee3"
      }
    }
  }
}


RESPONSE: [2020-04-15T09:32:35Z]
HTTP/1.1 201 Created
Content-Length: 1111
Content-Type: application/json; charset=utf-8
Date: Wed, 15 Apr 2020 09:32:35 GMT
Referrer-Policy: strict-origin-when-cross-origin
Server: nginx
Strict-Transport-Security: max-age=31536000; includeSubDomains; preload;
X-Content-Type-Options: nosniff
X-Download-Options: noopen
X-Frame-Options: SAMEORIGIN
X-Permitted-Cross-Domain-Policies: none
X-Ratelimit-Limit: 40000
X-Ratelimit-Remaining: 39954
X-Ratelimit-Reset: 1586946155
X-Runtime: 0.027783
X-Vcap-Request-Id: a2e67c03-73c6-4cd4-504f-0479dba0a0f3::2cc27d02-a4b5-4bc0-9bdd-d95e70deb47d
X-Xss-Protection: 1; mode=block
{
  "created_at": "2020-04-15T09:32:35Z",
  "destinations": null,
  "guid": "4cc15ddd-c360-4f89-8d84-52c97b4203d8",
  "host": "this_should_not_work",
  "links": {
    "destinations": {
      "href": "https://api.cf.mycloud.com/v3/routes/4cc15ddd-c360-4f89-8d84-52c97b4203d8/destinations"
    },
    "domain": {
      "href": "https://api.cf.mycloud.com/v3/domains/c542069e-06a6-481c-b8ba-a055bb2cd086"
    },
    "self": {
      "href": "https://api.cf.mycloud.com/v3/routes/4cc15ddd-c360-4f89-8d84-52c97b4203d8"
    },
    "space": {
      "href": "https://api.cf.mycloud.com/v3/spaces/c656ca01-6846-4f9d-a1e3-ff026ffaeee3"
    }
  },
  "metadata": {
    "annotations": {},
    "labels": {}
  },
  "path": "",
  "relationships": {
    "domain": {
      "data": {
        "guid": "c542069e-06a6-481c-b8ba-a055bb2cd086"
      }
    },
    "space": {
      "data": {
        "guid": "c656ca01-6846-4f9d-a1e3-ff026ffaeee3"
      }
    }
  },
  "updated_at": "2020-04-15T09:32:35Z",
  "url": "this_should_not_work.cfapps.mycloud.com"
}


OK
```

**Comment by sethboyles** [comment]
Reopening this to evaluate if this is desired behavior

**Comment by sethboyles** [comment]
Here is a similar issue where the resolution was to allow the lenience to not include a breaking change and because underscores are supported by many devices: https://github.com/cloudfoundry/cloud_controller_ng/issues/859

**Comment by domdom82** [comment]
This issue can be closed IMHO.
Why?
Even though we've run into this problem in production, the problem mostly arises with DNS and certificates.
So if you run into trouble usually depends on your DNS provider and CA. Most DNS providers silently accept underscores in host names, even though it is technically illegal.
When it comes to CAs, most of them will not sign a domain name that contains an illegal character, however this may not be a problem if you are using a wildcard certificate (e.g. `*.mydomain.com`) because the CA won't care if you are hosting a CF app called `my_app_with_underscores.mydomain.com` under that domain it signed. If a client library doesn't work to e.g. compute a URL based on a string containing underscores, then you can still change the host name by mapping another (legal) route. 
I agree this change cannot be done now that so many apps would break and without much benefit either.

So bottom line: I would have preferred for CF to be RFC-compliant but I get that the deed is done and we cannot break existing apps that would have functioned fine otherwise.
